### ISSUE TITLE
gogs should create the pull requests between the same branches by default

### ISSUE THREAD

**Issue by roidelapluie** [gogits/gogs#2111]
Repo A has following branches:
stable
dev
master

repo B is a fork. When creating a pull request from branch dev, it creates it against master by default. It should create the pull request against the dev branch if possible.

**Comment by Unknwon** [comment]
Thanks your feedback!

**Comment by nanoant** [comment]
This is very good idea. Also it would be great if Gogs was detecting the nearest base of feature branch for PR. Eg.

      * 12345 finished a         (feature/a)
    * | 12345 development cont   (origin/develop)
    | * 12345 working on a
    |/
    * 12345 development start   (develop)
    * 12345 stable commit       (master) (origin/master)

Now if I want to merge `feature/a` into `origin` repo, Gogs should suggest me `develop` too, not `master`. This is tricky, because `develop` and `origin/develop` has diverged, but this is common situation.

**Comment by roidelapluie** [comment]
nanoant: it could also look for the name:

develop/feature/a -> develop

**Comment by nanoant** [comment]
@roidelapluie Maybe, but the name can be misleading or wrong. It think it should/could just infer the base from both user and origin repo. As on example above, it can be inferred that `feature/a` base is `develop`.

**Comment by roidelapluie** [comment]
Indeed would be great especially on large repos where checking against the wrong branch first can take a very long time

**Comment by Unknwon** [comment]
So.... do we have a conclusion here?

**Comment by roidelapluie** [comment]
@Unknwon The best approach is to take the 'closest' branch

**Comment by Unknwon** [comment]
Hmm... how to define `closest`?...

**Comment by nanoant** [comment]
Agreed. See my example above.

**Comment by nanoant** [comment]
> Hmm... how to define closest?...

The closest base on my repo that exists on the origin repo too. So `develop` on my example, as it is closest branch on my repo that exist on origin (place I want to merge to).
NOTE: That my `develop` and `origin/develop` may diverge, but this shouldn't be a problem for algorithm.

**Comment by Unknwon** [comment]
@nanoant Your example looks good, but may take too much calculation?

**Comment by roidelapluie** [comment]
@Unknwon the branch that has the bigger number of commits in common with the PR we have?

 ~/dev/puppet $ git log --oneline origin/2.7.x..master | wc -l
11469
 ~/dev/puppet $ git log --oneline origin/3.x..master | wc -l
2902

**Comment by nanoant** [comment]
@Unknwon I don't think it is costly for calculation, the PRs are usually just few or max tens of commits. So you just iterate via refs and check if this ref exists on target repo too.

**Comment by roidelapluie** [comment]
@nanoant it is not relevant to comare to local repo because we do not always update all the branches

**Comment by roidelapluie** [comment]
@nanoant if you have small repo.

I use repo where diff between branches can be thousands of commits, and that is why we need to efficiently find the right branch

**Comment by Unknwon** [comment]
The best solution I can think of is what @roidelapluie posted, compare the number of commits. However, this does not work in all the cases.

The safest way is to check merge base, then compare the number of commits find the smallest one. This is what I mean costly for calculation. Every time you open the compare and pull page, it runs N*(branch num) times git execution to get this result.

**Comment by nanoant** [comment]
@roidelapluie I disagree, in your proposed solution `feature/a` would be merged to `feature/b`, not to `develop` for following example:

      * 12345 finished a        (feature/a)
    * | 12345 working on b      (origin/feature/b)
    | * 12345 working on a
    |/
    * 12345 development cont
    * 12345 development start   (develop) (origin/develop)
    * 12345 stable commit       (master) (origin/master)

**Comment by roidelapluie** [comment]
@Unknwon if you do not want to implement this, there is an alternative: Do not process the preview page if there are more than X commits (or put a small timeout) so we do not compute very big diffs

**Comment by roidelapluie** [comment]
something like

" There are more than 100 commits to show. Do you really want a preview or select another branch to compare to?"

**Comment by Unknwon** [comment]
Maybe I'm not clear.... it is not the problem with diff..

It is the problem that get the merge base of head branch and all branches in base repository...

**Comment by roidelapluie** [comment]
yes but the diff IS a problem also if it picks the wrong branch

**Comment by Unknwon** [comment]
But user decides which branch he/she should send pull request to, not Gogs...

**Comment by roidelapluie** [comment]
@Unknwon Before the pull request is created, here is what would be great (attached picture
![cb](https://cloud.githubusercontent.com/assets/291750/11784231/83104be4-a27b-11e5-83a9-9482f0d181c1.png)
)

**Comment by roidelapluie** [comment]
I hope that it is more clear now

**Comment by Unknwon** [comment]
Thanks for the screenshot...

Hmm, then we still have not had the algorithm to get `closest` branch...

**Comment by roidelapluie** [comment]
No, starting for the name would be great to start with.

**Comment by Unknwon** [comment]
> No, starting with the exact name would be great to start with.

OK... sounds cool!

**Comment by nanoant** [comment]
Here's an example implementation of @roidelapluie solution in bash:
~~~
#!/bin/bash
prname=$1
best=999999

while read sha1 reftype refname; do
	count=$(git rev-list --use-bitmap-index --count $refname..$prname)
	echo Trying $count $refname
	if [[ $count -lt $best || $count -eq $best && ${#refname} -lt ${#bestref} ]]; then
		best=$count
		bestref=$refname
	fi
done < <(git for-each-ref --sort=-committerdate refs/remotes)

echo Best $best $bestref
~~~

Please note that you should never use `git log --one-line | wc -l` to count commits, there is `git rev-list --count` that does that and it is approximately 2-3x quicker. Also `--use-bitmap-index` when packs are using bitmap index (Git 2.0 feature) then the lookup is even 4x faster.

**Comment by Unknwon** [comment]
@nanoant Thanks your code!

But I still think commits number isn't a very good indicator for how close of two branches.

**Comment by nanoant** [comment]
Yeah. That makes sense. Especially that proposed solution can be slow for certain huge repositories. I tried to find some good solution, but I haven't found any good one. Here are some references:
* http://stackoverflow.com/questions/3161204/find-the-parent-branch-of-a-git-branch
* http://git.661346.n2.nabble.com/Finding-the-name-of-the-parent-branch-td2244222.html

I think default PR branch settings is the way to go.

Oh btw. does Gogs support changing PR base for existing PR? (in opposite to GitHub)

**Comment by Unknwon** [comment]
- [ ] Match exact branch name or fallback to default PR branch, or fallback again to default base branch
- [ ] When diff files are larger then X, show notice before showing the diff
